docs/src/proposals/interest-bearing-tokens.md

```diff
@@ -0,0 +1,331 @@
+[token-list releases](https://github.com/solana-labs/token-list/releases/latest),
+downloaded automatically at startup.
+
+### Associated Token Program
+
+The Associated Token Program needs to support all token programs seamlessly.
+Currently, it performs the following sequence:
+
+* `SystemInstruction::Transfer` the required lamports to make the SPL holding account rent-exempt
+* `SystemInstruction::Allocate` space for the SPL holding account
+* `SystemInstruction::Assign` the holding account to the SPL token program
+* `InitializeHolding` on SPL token program
+
+This does not work for an opaque token program, because we do not know the size required
+from the outside. Conversely, if we allow a token program to take the lamports it wants,
+a malicious token program could take too much from users.
+
+To get around this, the Token Program interface has a new `CreateHolding`
+instruction, which only allocates space and assigns the account to the program.
+Once the space is allocated, the Associated Token Program transfers the
+required lamports, and returns an error if that number is too large. The process
+becomes:
+
+* call `CreateHolding` on the token program (allocate and assign)
+* calculate rent requirement based on the data size of holding account
+* `SystemInstruction::Transfer` the required lamports to make the holding account rent-exempt
+* call `InitializeHolding` on the token program
+
+### Web3 / Wallets
+
+To properly support i-tokens, wallets and applications will need to make the
+following changes:
+
+* get holdings: query for holdings owned by the public key, for all official
+token programs listed in the registry (not just `Tokenkeg...`)
+* get balances: avoid deserializing holding data, and instead use
+`getBalance` from RPC
+* transfer tokens: only use `ApproveChecked` and `TransferChecked` on the token
+program
+* create holding: avoid directly creating an account and initializing,
+and use the associated token account program instead
+
+### On-chain Programs
+
+To properly support i-tokens on-chain, programs will need to make these changes:
+
+* multiple token programs: all programs that support SPL Tokens must always
+include the program id of the token(s) holdings they're passing in.
+For example, token-swap needs to accept separate token programs for token A and B
+* hard-coded token program: avoid using `Tokenkeg...` directly, and delete `id()`
+the SPL token code
+* get balance: instead of deserializing the account data into an SPL holding, use
+the new wrapper to deserialize or call `GetHoldingBalance` instruction with the
+appropriate token program
+* get owner: same as above, but using the `GetHoldingOwner`
+* get supply: same as above, but using the `GetMintSupply`
+* get mint authority: same as above, but using the `GetMintAuthority`
+* get mint decimals: same as above, but using the `GetMintDecimals`
+* transfer: only use `TransferChecked`, which requires passing mints
```

Comment:
considering https://github.com/solana-labs/solana/pull/15927/files#r597540110, I think any programs hard-coding some notable spl-tokens should start to check the authentity of given holding account by checking (program_address, mint_address) tuple, not only mint_address.

Reply:
Absolutely, we've been trying to push people in that direction as much as possible, but having it done automatically in the code will be even better